### PR TITLE
Fix dtype promotion and MLX training materialization

### PR DESCRIPTION
**Pull request by workofart**
## Description

1. Fix scalar constants defaulting to fp32 when they are used with bf16/fp16 tensors. Previously, expressions like `x + 0`, `x * 0.5`, or dropout masks could promote low precision activations back to fp32. Now scalar constants and dropout masks preserve the tensor dtype.
2. Update the GPT-2 pre-training config based on the smaller Chinchilla-law ([paper](https://arxiv.org/abs/2203.15556))  training target.
3. Update SFT to use the same model architecture as the pre-trained checkpoint, switch to the `wiki_GPT2_6000` checkpoint, and adjust SFT steps/lr/checkpoint cadence around that model size.
4. Remove the recursive optimizer parameter traversal now that model parameters are flat. This makes gradient reset / optimizer step behavior simpler and easier to reason about.
5. Add explicit MLX materialization boundaries after non-step microbatches and optimizer updates. Without this, MLX can hold onto lazy backward/update graphs across microbatches and eventually materialize too much work at a later sync point. Previously, memory can spike (or even OOM) suddenly at the "materialize" steps, which is not obvious.
6. Add `tokens_per_s` to trainer reports so we can see training throughput directly in the normal logs.

## Tests

Added unit tests.

New pre-training run using a much smaller model based on the chinchilla scaling law ([paper](https://arxiv.org/abs/2203.15556))
```
(ml-by-hand) ➜  ml-by-hand git:(main) ✗ python -m examples.gpt_2
2026-04-26 14:06:15,425 - INFO - 178255102 characters in the entire dataset. Sample:

April

April (Apr.) is the fourth month of the year in the Julian and Gregorian calendars, and come
2026-04-26 14:06:15,425 - INFO - Loading the vocabulary from disk.
2026-04-26 14:06:15,430 - INFO - Found existing encoded data at 'training_data/bpe_12000_wiki_simple_encoded_data.npz', loading it instead of re-encoding.
2026-04-26 14:06:15,431 - INFO - Vocabulary size: 12260
2026-04-26 14:06:15,431 - INFO - Encoded data length: 47193131
Data length: len(train_data)=42473817 len(test_data)=4719314
2026-04-26 14:06:15,431 - INFO - No resume_epoch given; initializing model & optimizer from scratch.
2026-04-26 14:06:16,268 - INFO - Training GPT2 with 39.56M parameters.
2026-04-26 14:06:16,268 - INFO - Fit plan:
{'mode': 'steps',
 'target_step': 25000,
 'report_every_steps': 50,
 'global_batch_size': 32,
 'micro_batch_size': 8,
 'gradient_accumulation_steps': 4}
Training:   0%|▏                   2026-04-26 14:08:43,189 - INFO - [Step 50]                 | 50/25000 [02:24<19:37:57,  2.83s/it]

...

Training:  24%|██████████████████  2026-04-26 18:54:41,977 - INFO - Saved checkpoint to checkpoints/wiki_GPT2_6000.json and checkpoints/wiki_GPT2_6000.npz
2026-04-26 18:54:41,978 - INFO - [Step 6000]
{'step': 6000,
 'train_loss': '3.9651',
 'tokens_per_s': '11220.3798',
 'val_loss': '3.5655',
 'grad_l2_norm': '0.3601',
 'lr': '0.000958'}

...

> python scripts/inference.py checkpoints/wiki_GPT2_6000.npz 128
2026-04-26 21:24:34,709 - INFO - Loading the vocabulary from disk.
2026-04-26 21:24:34,716 - INFO - Model:

Spring is

Yunya Abis

Yunya Abis (born November 13, 1982) is a Nigerian politician. He is a member of the House of Representatives of the Nigeria National Guard.

Abidjan was born in Ibrahimaki, Nigeria. He studied at the University of Zano University and studied at the Kubaraya Gentino University of Zavira.

Abidjan was married to Gamba Abisiband from 2003 until 2001 and they had six children. He lives in Mangad, Nigeria, Kenya,
```

**SFT run using the pre-trained checkpoint**
```
(ml-by-hand) ➜  ml-by-hand git:(fix-dtype-bugs) ✗ python -m examples.sft_gpt_2
2026-04-26 20:34:06,446 - INFO - Loading the vocabulary from disk.
2026-04-26 20:34:06,450 - INFO - Attempting to load pretrained weights from checkpoint: /Users/henry/Git/ml-by-hand/checkpoints/wiki_GPT2_6000.json, /Users/henry/Git/ml-by-hand/checkpoints/wiki_GPT2_6000.npz
2026-04-26 20:34:06,484 - INFO - Loaded pretrained GPT2 weights from /Users/henry/Git/ml-by-hand/checkpoints/wiki_GPT2_6000
2026-04-26 20:34:06,519 - INFO - Found SFT cache at '/Users/henry/Git/ml-by-hand/training_data/bpe_12000_no_robots_encoded_data_sft.npz'; validating metadata before loading.
2026-04-26 20:34:06,524 - INFO - SFT cache metadata matched; loading cached tokenized data.
Data length: raw_train=9500 raw_val=500 fit_train=9248 fit_val=487
2026-04-26 20:34:06,974 - INFO - Training GPT2 with 39.56M parameters.
2026-04-26 20:34:06,974 - INFO - Fit plan:
{'mode': 'steps',
 'target_step': 900,
 'report_every_steps': 20,
 'global_batch_size': 32,
 'micro_batch_size': 8,
 'gradient_accumulation_steps': 4}
Training:   2%|█                  2026-04-26 20:34:48,477 - INFO - [Step 20]40<23:34,  1.61s/it]
{'step': 20,
 'train_loss': '4.9537',
 'tokens_per_s': '3220.3493',
 'val_loss': '4.1394',
 'grad_l2_norm': '1.0961',
 'lr': '0.000012'}

...


Training:  33%|███████████████▋    2026-04-26 20:41:50,897 - INFO - Saved checkpoint to checkpoints/sft_0426_GPT2_300.json and checkpoints/sft_0426_GPT2_300.npz
2026-04-26 20:41:50,897 - INFO - [Step 300]
{'step': 300,
 'train_loss': '4.7636',
 'tokens_per_s': '3946.2367',
 'val_loss': '3.8768',
 'grad_l2_norm': '0.7094',
 'lr': '0.000063'}

...

> python scripts/inference.py checkpoints/sft_0426_GPT2_300.npz 128
2026-04-26 20:43:18,692 - INFO - Loading the vocabulary from disk.
2026-04-26 20:43:18,702 - INFO - Model:

User: What is the weather today?<|endoftext|>Assistant:

The weather is usually very warm; the cold and moistest climates are hot and dry. It grows in hot weather or cold temperatures. Tropical Depression Twelve months later, in the day it gets cold to .

Sleep on Earth is about 30 degrees Celsius, but it becomes hot and dry.

Sleep on Earth is an average weather event. It is the average temperature recorded by the air. There are two different weather events:

### CODE REVIEW

File: examples/gpt_2.py [hunk 1]
```diff
@@ -262,20 +262,20 @@ def sample(self, model: GPT2, input_ids: Array) -> Tensor:
...
+        report_every_steps=50,
         global_batch_size=32,
-        micro_batch_size=4,
+        micro_batch_size=8,
         max_grad_norm=1.0,
         model_kwargs={
-            "num_attention_heads": 12,  # GPT-2 small uses 12
-            "hidden_size": 1536,  # GPT-2 small uses 768, must be divisible by num_attention_heads
-            "dropout_prob": 0.2,
+            "num_attention_heads": 9,  # GPT-2 small uses 12
```
Comment (workofart) [inline examples/gpt_2.py:273]: The updated model is about ~40M parameters.

The dataset we're using is https://www.kaggle.com/datasets/ffatty/plain-text-wikipedia-simpleenglish, which contains 31M tokens
number of tokens = 20 * number of parameters = 800M tokens
max seq len = 1024
global batch size = 32

tokens per step = 1024 * 32 = 32768
800M / 32768 = 24.4k steps

File: autograd/optim.py [hunk 2]
```diff
@@ -216,33 +216,6 @@ def update_lr(self) -> None:
         if self.lr_scheduler is not None:
             self.lr = self.lr_scheduler(self.timestep, self.initial_lr, self.lr)
 
-    def _recursive_param_op(
```
Comment (workofart) [inline autograd/optim.py:219]: This is actually not used anymore, because we already have a code comment:
https://github.com/workofart/ml-by-hand/blob/fedfed0a4fef588978d3fbd3ba1aa71fa48d0490/autograd/optim.py#L145-L146

I must have forgotten to deprecate this before

File: autograd/optim.py [hunk 3]
```diff
@@ -310,6 +281,14 @@ def grad_l2_norm(self) -> float:
                 grad_norm += (param.grad.data**2).sum()
         return float(xp.to_scalar(grad_norm**0.5))
 
+    def gradient_arrays(self) -> Dict[str, Array]:
```
Comment (workofart) [inline autograd/optim.py:284]: This is an external API that trainer can use to iterate over the model params without knowing the details of the `self.model_parameters` internals.

File: autograd/optim.py [hunk 4]
```diff
@@ -373,9 +352,6 @@ def step(self) -> None:
         """
         raise NotImplementedError
 
-    def _eval_backend(self) -> None:
```
Comment (workofart) [inline autograd/optim.py:376]: Removed this and rely on the `materialize` API provided by `backend.py`

File: autograd/tensor.py [hunk 5]
```diff
@@ -384,8 +385,7 @@ def __add__(self, other: Union["Tensor", float, int]) -> "Tensor":
         Returns:
             Tensor: The result of addition.
         """
-        if not isinstance(other, Tensor):
```
Comment (workofart) [inline autograd/tensor.py:387]: Previously, we were initializing the `other` (if scalar) into a Tensor with default dtype of FP32, regardless of what dtype `self` is. This would break the mixed precision training, and cause all downstream BF16 operations to output FP32 because the operation is performed between FP32 and BF16, and that will convert the output to FP32.

File: autograd/tensor.py [hunk 6]
```diff
@@ -468,14 +466,24 @@ def __iadd__(self, other: Union["Tensor", float, int]) -> "Tensor":
...
-            other = Tensor(other, requires_grad=False)
+        other = self._wrap_scalar_like_self(other)
 
         # Use expand for broadcasting
         broadcast_shape = xp.broadcast_shapes(self.shape, other.shape)
         expanded_other = other.expand(broadcast_shape)
         return IAdd.apply(self, expanded_other)
 
+    def _wrap_scalar_like_self(self, value: Union["Tensor", float, int]) -> "Tensor":
```
Comment (workofart) [inline autograd/tensor.py:476]: This helper function will help preserve whatever the low precision dtype (of `self`) when initializing the `other` tensor. If `self` is not low precision (e.g. softmax high-precision calculation), we still initialize `other` as high precision default of FP32.
